In [1]:
import pandas as pd
import numpy as np

columns = [
    "duration","protocol_type","service","flag","src_bytes","dst_bytes","land",
    "wrong_fragment","urgent","hot","num_failed_logins","logged_in","num_compromised",
    "root_shell","su_attempted","num_root","num_file_creations","num_shells",
    "num_access_files","num_outbound_cmds","is_host_login","is_guest_login","count",
    "srv_count","serror_rate","srv_serror_rate","rerror_rate","srv_rerror_rate",
    "same_srv_rate","diff_srv_rate","srv_diff_host_rate","dst_host_count",
    "dst_host_srv_count","dst_host_same_srv_rate","dst_host_diff_srv_rate",
    "dst_host_same_src_port_rate","dst_host_srv_diff_host_rate","dst_host_serror_rate",
    "dst_host_srv_serror_rate","dst_host_rerror_rate","dst_host_srv_rerror_rate",
    "label","difficulty"
]

df = pd.read_csv("../data/KDDTrain+.txt", names=columns)
df["binary_label"] = df["label"].apply(lambda x: "normal" if x == "normal" else "attack")

df.shape

(125973, 44)

In [2]:
df = df.drop(columns=["label", "difficulty"])
df.shape

(125973, 42)

In [3]:
for col in ["duration", "src_bytes", "dst_bytes"]:
    df[col] = np.log1p(df[col])

df[["duration", "src_bytes", "dst_bytes"]].describe()

,duration,src_bytes,dst_bytes
count,125973.000000,125973.000000,125973.000000
mean,0.321650,3.229669,3.084400
std,1.451379,2.982191,3.548927
min,0.000000,0.000000,0.000000
25%,0.000000,0.000000,0.000000
50%,0.000000,3.806662,0.000000
75%,0.000000,5.624018,6.248043
max,10.666837,21.045323,20.993245


In [4]:
df = pd.get_dummies(df, columns=["protocol_type", "service", "flag"], drop_first=True)
df.shape

(125973, 120)

In [5]:
df["binary_label"] = df["binary_label"].map({"normal": 0, "attack": 1})
df["binary_label"].value_counts()

binary_label
0    67343
1    58630
Name: count, dtype: int64

In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 125973 entries, 0 to 125972
Columns: 120 entries, duration to flag_SH
dtypes: bool(81), float64(18), int64(21)
memory usage: 47.2 MB


In [8]:
df.to_csv("../data/processed_train.csv", index=False)

## Summary

- Dropped `label` (redundant with binary_label) and `difficulty` (metadata, not a real feature).
- Applied log1p transformation to duration, src_bytes, dst_bytes to fix right-skew.
- One-hot encoded protocol_type, service, and flag (42 -> 120 columns).
- Encoded binary_label: normal=0, attack=1.
- Saved cleaned dataset to data/processed_train.csv for use in Day 5.